walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,151 @@
+   *                a name, a duration in minutes and a collection of attendees.
+   *                The attendees are either mandatory or optional.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Set<Event> eventsSet = new HashSet<>(events);
+
+    // Get the mandatory attendees of the new meeting.
+    Set<String> mandatoryAttendees = new HashSet<>(request.getAttendees());
+    
+    // Get both mandatory and optional attendees.
+    Set<String> allAttendees = new HashSet<>(request.getAttendees());
+    allAttendees.addAll(request.getOptionalAttendees());
+
+    List<TimeRange> timeRangesForAllAttendees = 
+        solveQueryforAttendees(eventsSet, allAttendees, request.getDuration());
+    
+    // If one or more time slots exists so that both mandatory and optional attendees can attend, 
+    // return those time slots. 
+    if (!timeRangesForAllAttendees.isEmpty()) {
+      return timeRangesForAllAttendees;
+    }
+    
+    // Otherwise, return the time slots that fit just the mandatory attendees.
+    return solveQueryforAttendees(eventsSet, mandatoryAttendees, request.getDuration());
+  }
+
+  /**
+   * Return the times when a new meeting could happen in a day, so that all {@code attendees} can join.
+   *
+   * @param events    The set of all known events in that day.
+   * @param attendees The set of required attendees for the new meeting.
+   * @param duration  The time duration of the new meeting.
+   */
+  private static List<TimeRange> solveQueryforAttendees(Set<Event> events, Set<String> attendees, long duration) {
+    // Get the time ranges when at least one attendee is not free 
+    // (because for a particular time slot to work, all attendees must be free to attend the new meeting).
+    List<TimeRange> occupiedTimeRanges = events
+      .stream()
+      .filter(event -> hasCommonAttendees(event, attendees))
+      .map(event -> event.getWhen())
+      .filter(timeRange -> checkIsWithinTheDay(timeRange))
+      .collect(Collectors.toList());
+    
+    return getNewMeetingTimeRanges(occupiedTimeRanges, duration);
+  }
+
+  /**
+   * Returns whether the {@code event} has at least one common attendee as the {@code attendees}.
+   */
+  private static boolean hasCommonAttendees(Event event, Set<String> attendees) {
+    Set<String> eventAttendees = new HashSet<>(event.getAttendees());
+    for (String attendee : attendees) {
+      if (eventAttendees.contains(attendee)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * Returns whether the {@code timeRange} is within the day.
+   */
+  private static boolean checkIsWithinTheDay(TimeRange timeRange) {
+    if (timeRange.WHOLE_DAY.contains(timeRange)) {
+      return true;
+    } else {
+      throw new IllegalArgumentException("Time ranges can only be within a day.");
```

Comment:
If I understand correctly, IAE will result in 500s. However, in this case, it's a client error and thus 400 is a better fit. But it's fine to keep it.

Reply:
Thank you for this notice, I will take it into consideration from now on!